nkhristinin/kibana
# Gaps update POC

## Event log changes:

- It support updating document right now. Because it's datastream, [we have to](https://www.elastic.co/guide/en/elasticsearch/reference/current/use-a-data-stream.html#update-delete-docs-in-a-backing-index) search for those document got get index, and other parameters and then make update. it's not a bulk, as we need to wait until operations is finished and retry (TODO). Also main case will be that there only 1 gap.
- Support documents removing by query


## Gap updates

- After backfill is scheduled, we find all gaps in in this date range, and add those intervasl into in_progress intervals
- After rule executed, we find all gaps and add date range into fileld intervals
- If rule failed or backfill is deleted, we try to recalculate in progress gap. We find all gaps in this date range, then for each gap we find all backfills which is scheduled and we add calculate in progress intervals
- If rule is removed - all gaps are removed 

## How to test

Enable feature flag `storeGapsInEventLogEnabled`
Create security rule - (1m interval 1s lookback time)
Enable rule, wait for  successful execution
Disable rule, wait 5 minutes, enable rule and observe rule failure with gap message

Go to rule executions tab, and see the gap table. It's possible to fill the gap.
